# Spatially variable genes

Spatially variable genes (SVGs) are genes whose expression levels vary significantly across different spatial regions within a tissue or across cells in a spatially structured context.

Path:
[`src/tasks/spatially_variable_genes`](https://github.com/openproblems-bio/openproblems-v2/tree/main/src/tasks/spatially_variable_genes)

## Motivation

Recent years have witnessed significant progress in spatially-resolved transcriptome profiling techniques that simultaneously characterize cellular gene expression and their physical position, generating spatial transcriptomic (ST) data. The application of these techniques has dramatically advanced our understanding of disease and developmental biology. One common task for all ST profiles, regardless of the employed protocols, is to identify genes that exhibit spatial patterns. These genes, defined as spatially variable genes (SVGs), contain additional information about the spatial structure of the tissues of interest, compared to highly variable genes (HVGs).

## Description

<!-- We distinguish between *reference-based* decomposition and *de novo*
decomposition, where the former leverage external data (e.g., scRNA-seq
or scNuc-seq) to guide the inference process, while the latter only work
with the spatial data. We require that all datasets have an associated
reference single cell data set, but methods are free to ignore this
information. Due to the lack of real datasets with the necessary
ground-truth, this task makes use of a simulated dataset generated by
creating cell-aggregates by sampling from a Dirichlet distribution. The
ground-truth dataset consists of the spatial expression matrix, XY
coordinates of the spots, true cell-type proportions for each spot, and
the reference single-cell data (from which cell aggregated were
simulated). -->

## Authors & contributors

| name             | roles              |
|:-----------------|:-------------------|
| Zhijian Li   | author, maintainer |
| Zain M. Patel| author             |

## API

``` mermaid
flowchart LR
  file_common_dataset("Common Dataset")
  comp_process_dataset[/"Data processor"/]
  file_single_cell("Single cell data")
  file_spatial_masked("Spatial masked")
  file_solution("Solution")
  comp_control_method[/"Control method"/]
  comp_method[/"Method"/]
  comp_metric[/"Metric"/]
  file_output("Output")
  file_score("Score")
  file_common_dataset---comp_process_dataset
  comp_process_dataset-->file_single_cell
  comp_process_dataset-->file_spatial_masked
  comp_process_dataset-->file_solution
  file_single_cell---comp_control_method
  file_single_cell---comp_method
  file_spatial_masked---comp_control_method
  file_spatial_masked---comp_method
  file_solution---comp_control_method
  file_solution---comp_metric
  comp_control_method-->file_output
  comp_method-->file_output
  comp_metric-->file_score
  file_output---comp_metric
```

## File format: Common Dataset

A subset of the common dataset.

Example file:
`resources_test/spatially_variable_genes/10x_Visium_mouse_brain/dataset_simulated.h5ad`

Format:

<div class="small">
    AnnData object
     var: 'spatial_var'
     obsm: 'coordinates''
     layers: 'counts'
     uns: 'dataset_id', 'dataset_name', 'dataset_url', 'dataset_reference', 'dataset_summary', 'dataset_description', 'dataset_organism'
</div>

Slot description:

<div class="small">

| Slot                          | Type      | Description |
|:------------------------------|:----------|:-------------|
| `var["svg"]`                  | `boolean` | Whether or not the feature is considered to be a ‘highly variable gene’.|
| `var["spatial_var"]`          | `double`  | A ranking of the features by svg. |
| `obsm["coordinates"]`         | `double`  | XY coordinates for each spot. |
| `layers["counts"]`            | `integer` | Raw counts.|
| `uns["dataset_id"]`           | `string`  | A unique identifier for the dataset. |
| `uns["dataset_name"]`         | `string`  | Nicely formatted name.|
| `uns["dataset_url"]`          | `string`  | (*Optional*) Link to the original source of the dataset. |
| `uns["dataset_reference"]`    | `string`  | (*Optional*) Bibtex reference of the paper in which the dataset was published.|
| `uns["dataset_summary"]`      | `string`  | Short description of the dataset.|
| `uns["dataset_description"]`  | `string`  | Long description of the dataset.|
| `uns["dataset_organism"]`     | `string`  | (*Optional*) The organism of the sample in the dataset.|

</div>

## Component type: Data processor

Path:
[`src/spatially_variable_genes`](https://github.com/openproblems-bio/openproblems-v2/tree/main/src/spatially_variable_genes)

A spatially variable gene dataset processor.

Arguments:

<div class="small">

| Name          | Type   | Description |
|:--------------|:-------|:------------|
| `--input`              | `file` | A subset of the common dataset. |
| `--output_solution`    | `file` | The data for evaluating spatially variable gene detection. |

</div>

## File format: Spatial data

The spatial data file containing transcription profiles for each capture
location.

Example file:
`resources_test/spatially_variable_genes/cxg_mouse_pancreas_atlas/spatial_masked.h5ad`

Format:

<div class="small">

    AnnData object
     obsm: 'coordinates'
     layers: 'counts'
     uns: 'dataset_id'

</div>

Slot description:

<div class="small">

| Slot                     | Type      | Description   |
|:-------------------------|:----------|:--------------|
| `obsm["coordinates"]`    | `double`  | XY coordinates for each spot. |
| `layers["counts"]`       | `integer` | Raw counts.   |
| `uns["dataset_id"]`      | `string`  | A unique identifier for the dataset. |

</div>

## File format: Solution

A CSV file containing estimated spatial variability for each gene.

Example file:
`resources_test/spatially_variable_genes/10x_Visium_mouse_brain/solution.csv`

Format:

<div class="small">

    AnnData object
     obsm: 'coordinates', 'proportions_true'
     layers: 'counts'
     uns: 'cell_type_names', 'dataset_id'

</div>

Slot description:

<div class="small">

| Slot                       | Type      | Description                                                |
|:---------------------------|:----------|:-----------------------------------------------------------|
| `obsm["coordinates"]`      | `double`  | XY coordinates for each spot.                              |
| `obsm["proportions_true"]` | `double`  | True cell type proportions for each spot.                  |
| `layers["counts"]`         | `integer` | Raw counts.                                                |
| `uns["cell_type_names"]`   | `string`  | Cell type names corresponding to columns of `proportions`. |
| `uns["dataset_id"]`        | `string`  | A unique identifier for the dataset.                       |

</div>

## Component type: Control method

Path:
[`src/spatially_variable_genes/control_methods`](https://github.com/openproblems-bio/openproblems-v2/tree/main/src/spatially_variable_genes/control_methods)

Quality control methods for verifying the pipeline.

Arguments:

<div class="small">

| Name                     | Type   | Description                                                                                                                                          |
|:-------------------------|:-------|:-----------------------------------------------------------------------------------------------------------------------------------------------------|
| `--input` | `file` | The spatial data file containing transcription profiles for each capture location.  |
| `--output`   | `file` | (*Output*) CSV file with estimated spatial variability. |

</div>

## Component type: Method

Path:
[`src/spatially_variable_genes/methods`](https://github.com/openproblems-bio/openproblems-v2/tree/main/src/spatially_variable_genes/methods)

A spatially variable gene detection method.

Arguments:

<div class="small">

| Name                     | Type   | Description                                                                                                                     |
|:-------------------------|:-------|:--------------------------------------------------------------------------------------------------------------------------------|
| `--input` | `file` | The spatial data file containing transcription profiles for each capture location.  |
| `--output`   | `file` | (*Output*) CSV file with estimated spatial variability. |

</div>

## Component type: Metric

Path:
[`src/spatial_decomposition/metrics`](https://github.com/openproblems-bio/openproblems-v2/tree/main/src/spatial_decomposition/metrics)

A spatial decomposition metric.

Arguments:

<div class="small">

| Name               | Type   | Description                                                                                                                                          |
|:-------------------|:-------|:-----------------------------------------------------------------------------------------------------------------------------------------------------|
| `--input_method`   | `file` | Spatial data with estimated proportions.                                                                                                             |
| `--input_solution` | `file` | The spatial data file containing transcription profiles for each capture location, with true cell-type proportions for each spot / capture location. |
| `--output`         | `file` | (*Output*) Metric score file.                                                                                                                        |

</div>

## File format: Output

CSV file

Spatial data with estimated proportions.

Example file:
`resources_test/spatially_variable_genes/cxg_mouse_pancreas_atlas/output.h5ad`

Description:

Spatial data file with estimated cell type proportions.

Format:

<div class="small">

    AnnData object
     obsm: 'coordinates', 'proportions_pred'
     layers: 'counts'
     uns: 'cell_type_names', 'dataset_id', 'method_id'

</div>

Slot description:

<div class="small">

| Slot                       | Type      | Description                                                |
|:---------------------------|:----------|:-----------------------------------------------------------|
| `obsm["coordinates"]`      | `double`  | XY coordinates for each spot.                              |
| `obsm["proportions_pred"]` | `double`  | Estimated cell type proportions for each spot.             |
| `layers["counts"]`         | `integer` | Raw counts.                                                |
| `uns["cell_type_names"]`   | `string`  | Cell type names corresponding to columns of `proportions`. |
| `uns["dataset_id"]`        | `string`  | A unique identifier for the dataset.                       |
| `uns["method_id"]`         | `string`  | A unique identifier for the method.                        |

</div>

## File format: Score

Metric score file.

Example file:
`resources_test/spatial_decomposition/cxg_mouse_pancreas_atlas/score.h5ad`

Format:

<div class="small">

    AnnData object
     uns: 'dataset_id', 'method_id', 'metric_ids', 'metric_values'

</div>

Slot description:

<div class="small">

| Slot                   | Type     | Description                                                                                  |
|:-----------------------|:---------|:---------------------------------------------------------------------------------------------|
| `uns["dataset_id"]`    | `string` | A unique identifier for the dataset.                                                         |
| `uns["method_id"]`     | `string` | A unique identifier for the method.                                                          |
| `uns["metric_ids"]`    | `string` | One or more unique metric identifiers.                                                       |
| `uns["metric_values"]` | `double` | The metric values obtained for the given prediction. Must be of same length as ‘metric_ids’. |

</div>

